The job **`mingw-check-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15427728292/job/43418899323) [(plain)](https://github.com/rust-lang/rust/commit/b98674605f38be31652dea13093abd3e3700cf1e/checks/43418899323/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[RUSTC-TIMING] rustc_ast_lowering test:true 1.285
[RUSTC-TIMING] rustc_metadata test:true 2.081
[RUSTC-TIMING] rustc_ast_ir test:true 0.072
[RUSTC-TIMING] rustc_smir test:true 1.644
error[E0277]: the trait bound `i32: idx::Idx` is not satisfied
   --> compiler/rustc_index/src/bit_set/tests.rs:697:19
    |
697 |     let mut set = DenseBitSet::new_empty(300);
    |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `idx::Idx` is not implemented for `i32`
    |
    = help: the following other types implement trait `idx::Idx`:
              u32
              usize
note: required by a bound in `bit_set::DenseBitSet::<T>::new_empty`
   --> compiler/rustc_index/src/bit_set.rs:132:9
    |
132 | impl<T: Idx> DenseBitSet<T> {
    |         ^^^ required by this bound in `DenseBitSet::<T>::new_empty`
...
135 |     pub fn new_empty(domain_size: usize) -> DenseBitSet<T> {
    |            --------- required by a bound in this associated function

error[E0277]: the trait bound `i32: idx::Idx` is not satisfied
   --> compiler/rustc_index/src/bit_set/tests.rs:698:18
    |
698 |     assert!(!set.contains_any(0..300));
    |                  ^^^^^^^^^^^^ the trait `idx::Idx` is not implemented for `i32`
    |
    = help: the following other types implement trait `idx::Idx`:
              u32
              usize
note: required by a bound in `bit_set::DenseBitSet::<T>::contains_any`
   --> compiler/rustc_index/src/bit_set.rs:132:9
    |
132 | impl<T: Idx> DenseBitSet<T> {
    |         ^^^ required by this bound in `DenseBitSet::<T>::contains_any`
...
239 |     pub fn contains_any(&self, elems: impl RangeBounds<T>) -> bool {
    |            ------------ required by a bound in this associated function

error[E0277]: the trait bound `i32: idx::Idx` is not satisfied
   --> compiler/rustc_index/src/bit_set/tests.rs:699:9
    |
699 |     set.insert_range(10..20);
    |         ^^^^^^^^^^^^ the trait `idx::Idx` is not implemented for `i32`
    |
    = help: the following other types implement trait `idx::Idx`:
              u32
              usize
note: required by a bound in `bit_set::DenseBitSet::<T>::insert_range`
   --> compiler/rustc_index/src/bit_set.rs:132:9
    |
132 | impl<T: Idx> DenseBitSet<T> {
    |         ^^^ required by this bound in `DenseBitSet::<T>::insert_range`
...
205 |     pub fn insert_range(&mut self, elems: impl RangeBounds<T>) {
    |            ------------ required by a bound in this associated function

For more information about this error, try `rustc --explain E0277`.
[RUSTC-TIMING] rustc_index test:true 0.425
error: could not compile `rustc_index` (lib test) due to 3 previous errors
```

</details>

--- a/compiler/rustc_index/src/bit_set.rs
+++ b/compiler/rustc_index/src/bit_set.rs
@@ -234,6 +234,32 @@ impl<T: Idx> DenseBitSet<T> {
         self.clear_excess_bits();
     }
 
+    /// Checks whether any bit in the given range is a 1.
+    #[inline]
+    pub fn contains_any(&self, elems: impl RangeBounds<T>) -> bool {
+        let Some((start, end)) = inclusive_start_end(elems, self.domain_size) else {
+            return false;
+        };
+        let (start_word_index, start_mask) = word_index_and_mask(start);
+        let (end_word_index, end_mask) = word_index_and_mask(end);
+
+        if start_word_index == end_word_index {
+            self.words[start_word_index] & (end_mask | (end_mask - start_mask)) != 0
+        } else {
+            if self.words[start_word_index] & !(start_mask - 1) != 0 {
+                return true;
+            }
+
+            let remaining = start_word_index + 1..end_word_index;
+            if remaining.start <= remaining.end {
+                self.words[remaining].iter().any(|&w| w != 0)
+                    || self.words[end_word_index] & (end_mask | (end_mask - 1)) != 0
+            } else {
+                false
+            }
+        }
+    }
+
     /// Returns `true` if the set has changed.
     #[inline]
     pub fn remove(&mut self, elem: T) -> bool {

--- a/compiler/rustc_index/src/bit_set/tests.rs
+++ b/compiler/rustc_index/src/bit_set/tests.rs
@@ -692,6 +692,25 @@ fn dense_last_set_before() {
     }
 }
 
+#[test]
+fn dense_contains_any() {
+    let mut set: DenseBitSet<usize> = DenseBitSet::new_empty(300);
+    assert!(!set.contains_any(0..300));
+    set.insert_range(10..20);
+    set.insert_range(60..70);
+    set.insert_range(150..=250);
+
+    assert!(set.contains_any(0..30));
+    assert!(set.contains_any(5..100));
+    assert!(set.contains_any(250..255));
+
+    assert!(!set.contains_any(20..59));
+    assert!(!set.contains_any(256..290));
+
+    set.insert(22);
+    assert!(set.contains_any(20..59));
+}
+
 #[bench]
 fn bench_insert(b: &mut Bencher) {
     let mut bs = DenseBitSet::new_filled(99999usize);

--- a/src/tools/miri/src/alloc/isolated_alloc.rs
+++ b/src/tools/miri/src/alloc/isolated_alloc.rs
@@ -145,10 +145,7 @@ impl IsolatedAlloc {
             if pinfo.domain_size() < offset_pinfo + size_pinfo {
                 break;
             }
-            // FIXME: is there a more efficient way to check whether the entire range is unset
-            // in the bitset?
-            let range_avail = !(offset_pinfo..offset_pinfo + size_pinfo).any(|i| pinfo.contains(i));
-            if range_avail {
+            if !pinfo.contains_any(offset_pinfo..offset_pinfo + size_pinfo) {
                 pinfo.insert_range(offset_pinfo..offset_pinfo + size_pinfo);
                 // SAFETY: We checked the available bytes after `idx` in the call
                 // to `domain_size` above and asserted there are at least `idx +

